File: src/components/survival/tutorial/survivalTutorialDemoPlayScheduler.ts
import type {
  SurvivalTutorialV3DemoChordEvent,
  SurvivalTutorialV3DemoLine,
  SurvivalTutorialV3DemoPlayScene,
} from '@/components/survival/tutorial/survivalTutorialV3ScriptTypes';
import { SURVIVAL_TUTORIAL_V3_DIALOGUE_LINE_SECONDS } from '@/components/survival/tutorial/survivalTutorialV3Constants';

export interface SurvivalTutorialDemoScheduleEvent {
  readonly kind: 'chord-start' | 'roll-step-start' | 'chord-end' | 'line-start' | 'line-end' | 'demo-end';
  readonly atBeat: number;
  readonly atSeconds: number;
  readonly chordIndex?: number;
  readonly rollStepIndex?: number;
  readonly lineIndex?: number;
}

export const beatToSeconds = (beat: number, bpm: number): number => {
  const safeBpm = Math.max(1, bpm);
  return (beat * 60) / safeBpm;
};

/** BGM アンカー時刻からの残り遅延 ms（負値は 0 にクランプ）。 */
export const anchoredDelayMs = (atSecondsMs: number, elapsedMs: number): number =>
  Math.max(0, atSecondsMs - elapsedMs);

export const measureForBeat = (beat: number, beatsPerMeasure: number): number =>
  Math.floor(beat / beatsPerMeasure) + 1;

export const defaultLineDurationBeats = (bpm: number): number =>
  (SURVIVAL_TUTORIAL_V3_DIALOGUE_LINE_SECONDS * bpm) / 60;

export const buildDemoPlaySchedule = (
  scene: SurvivalTutorialV3DemoPlayScene,
): readonly SurvivalTutorialDemoScheduleEvent[] => {
  const bpm = scene.bpm;
  const beatsPerMeasure = scene.beatsPerMeasure ?? 4;
  const defaultLineBeats = defaultLineDurationBeats(bpm);
  const endHold = scene.endHoldBeats ?? beatsPerMeasure;

  let maxEndBeat = 0;

  const events: SurvivalTutorialDemoScheduleEvent[] = [];

  scene.chords.forEach((chord, chordIndex) => {
    const rollSteps = chord.rollSteps;
    if (rollSteps && rollSteps.length > 0) {
      const end = chord.startBeat + chord.durationBeats;
      maxEndBeat = Math.max(maxEndBeat, end);
      rollSteps.forEach((step, rollStepIndex) => {
        const kind = rollStepIndex === 0 ? 'chord-start' : 'roll-step-start';
        events.push({
          kind,
          atBeat: step.startBeat,
          atSeconds: beatToSeconds(step.startBeat, bpm),
          chordIndex,
          rollStepIndex,
        });
      });
      events.push({
        kind: 'chord-end',
        atBeat: end,
        atSeconds: beatToSeconds(end, bpm),
        chordIndex,
      });
      return;
    }

    const start = chord.startBeat;
    const end = start + chord.durationBeats;
    maxEndBeat = Math.max(maxEndBeat, end);
    events.push({
      kind: 'chord-start',
      atBeat: start,
      atSeconds: beatToSeconds(start, bpm),
      chordIndex,
    });
    events.push({
      kind: 'chord-end',
      atBeat: end,
      atSeconds: beatToSeconds(end, bpm),
      chordIndex,
    });
  });

  scene.lines.forEach((line, lineIndex) => {
    const start = line.startBeat;
    const duration = line.durationBeats ?? defaultLineBeats;
    const end = start + duration;
    maxEndBeat = Math.max(maxEndBeat, end);
    events.push({
      kind: 'line-start',
      atBeat: start,
      atSeconds: beatToSeconds(start, bpm),
      lineIndex,
    });
    events.push({
      kind: 'line-end',
      atBeat: end,
      atSeconds: beatToSeconds(end, bpm),
      lineIndex,
    });
  });

  const demoEndBeat = maxEndBeat + endHold;
  events.push({
    kind: 'demo-end',
    atBeat: demoEndBeat,
    atSeconds: beatToSeconds(demoEndBeat, bpm),
  });

  events.sort((a, b) => a.atSeconds - b.atSeconds || a.kind.localeCompare(b.kind));
  return events;
};

export const chordEventId = (chord: SurvivalTutorialV3DemoChordEvent, index: number): string =>
  `demo-chord:${index}:${chord.measureNumber}:${chord.startBeat}`;

export const resolveDemoLineSpeaker = (
  line: SurvivalTutorialV3DemoLine,
): 'fai' | 'jajii' | 'narration' => {
  if (line.speaker === 'fai' || line.speaker === 'jajii' || line.speaker === 'narration') {
    return line.speaker;
  }
  return 'jajii';
};

export const getDemoPlayTotalBeats = (scene: SurvivalTutorialV3DemoPlayScene): number => {
  const schedule = buildDemoPlaySchedule(scene);
  const end = schedule.find((e) => e.kind === 'demo-end');
  return end?.atBeat ?? 0;
};

File: ios/Jazzify/Tutorial/SurvivalTutorial/SurvivalTutorialDemoPlayScheduler.swift
import Foundation

enum SurvivalTutorialDemoPlayScheduler {
    enum EventKind: String, Sendable {
        case chordStart
        case rollStepStart
        case chordEnd
        case lineStart
        case lineEnd
        case demoEnd
    }

    struct ScheduleEvent: Sendable {
        let kind: EventKind
        let atBeat: Double
        let atSeconds: Double
        let chordIndex: Int?
        let rollStepIndex: Int?
        let lineIndex: Int?
    }

    static func beatToSeconds(beat: Double, bpm: Double) -> Double {
        let safeBpm = max(1, bpm)
        return (beat * 60) / safeBpm
    }

    /// BGM アンカー時刻からの残り遅延秒（負値は 0 にクランプ）。
    /// `advanceSeconds` 正値で発火を前倒し（描画レイテンシ補正）、負値で後ろ倒し（出力レイテンシ追従）。
    static func anchoredDelaySeconds(
        atSeconds: Double,
        elapsedSeconds: Double,
        advanceSeconds: Double = 0
    ) -> Double {
        max(0, atSeconds - elapsedSeconds - advanceSeconds)
    }

    static func defaultLineDurationBeats(bpm: Double) -> Double {
        (SurvivalTutorialV3Constants.dialogueLineSeconds * bpm) / 60
    }

    static func buildSchedule(scene: SurvivalTutorialV3DemoPlayScene) -> [ScheduleEvent] {
        let bpm = scene.bpm
        let beatsPerMeasure = scene.beatsPerMeasure ?? 4
        let defaultLineBeats = defaultLineDurationBeats(bpm: bpm)
        let endHold = scene.endHoldBeats ?? beatsPerMeasure
        var maxEndBeat: Double = 0
        var events: [ScheduleEvent] = []

        for (index, chord) in scene.chords.enumerated() {
            if let rollSteps = chord.rollSteps, !rollSteps.isEmpty {
                let end = chord.startBeat + chord.durationBeats
                maxEndBeat = max(maxEndBeat, end)
                for (rollStepIndex, step) in rollSteps.enumerated() {
                    let kind: EventKind = rollStepIndex == 0 ? .chordStart : .rollStepStart
                    events.append(
                        ScheduleEvent(
                            kind: kind,
                            atBeat: step.startBeat,
                            atSeconds: beatToSeconds(beat: step.startBeat, bpm: bpm),
                            chordIndex: index,
                            rollStepIndex: rollStepIndex,
                            lineIndex: nil
                        )
                    )
                }
                events.append(
                    ScheduleEvent(
                        kind: .chordEnd,
                        atBeat: end,
                        atSeconds: beatToSeconds(beat: end, bpm: bpm),
                        chordIndex: index,
                        rollStepIndex: nil,
                        lineIndex: nil
                    )
                )
                continue
            }

            let start = chord.startBeat
            let end = start + chord.durationBeats
            maxEndBeat = max(maxEndBeat, end)
            events.append(
                ScheduleEvent(
                    kind: .chordStart,
                    atBeat: start,
                    atSeconds: beatToSeconds(beat: start, bpm: bpm),
                    chordIndex: index,
                    rollStepIndex: nil,
                    lineIndex: nil
                )
            )
            events.append(
                ScheduleEvent(
                    kind: .chordEnd,
                    atBeat: end,
                    atSeconds: beatToSeconds(beat: end, bpm: bpm),
                    chordIndex: index,
                    rollStepIndex: nil,
                    lineIndex: nil
                )
            )
        }

        for (index, line) in scene.lines.enumerated() {
            let start = line.startBeat
            let duration = line.durationBeats ?? defaultLineBeats
            let end = start + duration
            maxEndBeat = max(maxEndBeat, end)
            events.append(
                ScheduleEvent(
                    kind: .lineStart,
                    atBeat: start,
                    atSeconds: beatToSeconds(beat: start, bpm: bpm),
                    chordIndex: nil,
                    rollStepIndex: nil,
                    lineIndex: index
                )
            )
            events.append(
                ScheduleEvent(
                    kind: .lineEnd,
                    atBeat: end,
                    atSeconds: beatToSeconds(beat: end, bpm: bpm),
                    chordIndex: nil,
                    rollStepIndex: nil,
                    lineIndex: index
                )
            )
        }

        let demoEndBeat = maxEndBeat + endHold
        events.append(
            ScheduleEvent(
                kind: .demoEnd,
                atBeat: demoEndBeat,
                atSeconds: beatToSeconds(beat: demoEndBeat, bpm: bpm),
                chordIndex: nil,
                rollStepIndex: nil,
                lineIndex: nil
            )
        )

        return events.sorted {
            if $0.atSeconds == $1.atSeconds {
                return $0.kind.rawValue < $1.kind.rawValue
            }
            return $0.atSeconds < $1.atSeconds
        }
    }

    static func resolveLineSpeaker(_ line: SurvivalTutorialV3DemoLine) -> SurvivalTutorialV3DialogueSpeaker {
        switch line.speaker {
        case "fai":
            return .fai
        case "narration":
            return .narration
        default:
            return .jajii
        }
    }

    static func resolveWindowStartMeasure(
        chords: [SurvivalTutorialV3DemoChordEvent],
        activeChordIndex: Int?
    ) -> Int {
        if let activeChordIndex,
           chords.indices.contains(activeChordIndex) {
            return chords[activeChordIndex].measure_number
        }
        return chords.first?.measure_number ?? 1
    }

    /// demo_play 台本の全 `voicing` から最大 MIDI（休符 `[]` は除外）。鍵盤スクロール用。
    static func maxVoicingMidi(in chords: [SurvivalTutorialV3DemoChordEvent]) -> Int? {
        var maxValue: Int?
        for chord in chords {
            for midi in chord.voicing {
                if maxValue == nil || midi > maxValue! {
                    maxValue = midi
                }
            }
        }
        return maxValue
    }

    /// `applyTutorialSceneKeyboardScroll` 用。Onboarding ii-V-I ではなく台本 voicing のみから構築する。
    static func resolvedChordsForKeyboardScroll(
        in chords: [SurvivalTutorialV3DemoChordEvent]
    ) -> [SurvivalResolvedChord] {
        chords.enumerated().compactMap { index, event in
            guard !event.voicing.isEmpty else { return nil }
            let names = SurvivalTutorialDemoStaffBuilder.voicingNames(for: event)
            return SurvivalResolvedChord.fromExplicitTutorialVoicing(
                id: "demo-play-scroll:\(index)",
                name: event.chordName,
                voicing: event.voicing,
                voicingNames: names,
                keyFifths: event.keyFifths ?? 0
            )
        }
    }
}

struct SurvivalTutorialDemoStaffSnapshot: Equatable {
    let chords: [SurvivalTutorialV3DemoChordEvent]
    let activeChordIndex: Int?
    let activeRollStepIndex: Int?
    let keyFifths: Int
    let windowStartMeasure: Int
}

enum SurvivalTutorialDemoStaffBuilder {
    static func voicingNames(for chord: SurvivalTutorialV3DemoChordEvent) -> [String] {
        if let names = chord.voicingNames, names.count == chord.voicing.count {
            return names
        }
        return chord.voicing.map { "M\($0)" }
    }

    static func voicingStaves(for chord: SurvivalTutorialV3DemoChordEvent, names: [String]) -> [Int] {
        if let staves = chord.voicing_staves, staves.count == names.count {
            return staves
        }
        return names.map { _ in 2 }
    }

    static func fixedStaves(for chords: [SurvivalTutorialV3DemoChordEvent]) -> [Int] {
        var staves = Set<Int>()
        for chord in chords {
            if let voicingStaves = chord.voicing_staves {
                for staff in voicingStaves where staff == 1 || staff == 2 {
                    staves.insert(staff)
                }
            }
            if let rollSteps = chord.rollSteps {
                for step in rollSteps {
                    if let stepStaves = step.voicing_staves {
                        for staff in stepStaves where staff == 1 || staff == 2 {
                            staves.insert(staff)
                        }
                    }
                }
            }
        }
        if staves.contains(1), staves.contains(2) {
            return [1, 2]
        }
        if staves.contains(1) {
            return [1]
        }
        if staves.contains(2) {
            return [2]
        }
        return [1, 2]
    }
}
